client/src/instance_spec/mod.rs

```diff
@@ -0,0 +1,224 @@
+
+    /// The PCI path at which to attach this device.
+    pub pci_path: PciPath,
+}
+
+//
+// Network devices.
+//
+// Because there is only one kind of virtual networking device (a virtio-net
+// device with an enlightened vNIC), networking devices don't have the same
+// BackendKind and DeviceKind enums that storage devices do, but they can
+// be added if needed in future versions.
+
+/// A network backend, specifically a virtual NIC on the host system for which
+/// Virtio network adapter integration is enabled.
+#[derive(Clone, Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct NetworkBackend {
+    /// The name of the vNIC to connect to.
+    pub vnic_name: String,
+}
+
+/// A virtual network adapter that presents a virtio network device interface.
+#[derive(Clone, Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct NetworkDevice {
+    /// The name of the device's backend.
+    pub backend_name: String,
+
+    /// The PCI path at which to attach this device.
+    pub pci_path: PciPath,
+}
+
+//
+// Serial ports.
+//
+
+/// A serial port identifier, which determines what I/O ports a guest can use to
+/// access a port.
+#[derive(Clone, Copy, Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub enum SerialPortNumber {
+    Com1,
+    Com2,
+    Com3,
+    Com4,
+}
+
+/// A serial port device.
+#[derive(Clone, Copy, Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct SerialPort {
+    /// The serial port number for this port.
+    pub num: SerialPortNumber,
+
+    /// The initial discard discipline for this port. If true, the device will
+    /// not buffer bytes written from the guest until the port's discipline
+    /// changes.
+    pub autodiscard: bool,
+}
+
+/// A PCI-PCI bridge.
+#[derive(Clone, Copy, Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct PciPciBridge {
+    /// The logical bus number of this bridge's downstream bus. Other devices
+    /// may use this bus number in their PCI paths to indicate they should be
+    /// attached to this bridge's bus.
+    pub downstream_bus: u8,
+
+    /// The PCI path at which to attach this bridge.
+    pub pci_path: PciPath,
+}
+
+/// A full instance specification. See the documentation for individual
+/// elements for more information about the fields in this structure.
+///
+/// Named devices and backends are stored in maps with object names as keys
+/// and devices/backends as values.
+#[derive(Clone, Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct InstanceSpec {
+    pub board: Board,
+
+    pub storage_devices: BTreeMap<String, StorageDevice>,
+    pub storage_backends: BTreeMap<String, StorageBackend>,
+    pub network_devices: BTreeMap<String, NetworkDevice>,
+    pub network_backends: BTreeMap<String, NetworkBackend>,
+    pub serial_ports: BTreeMap<String, SerialPort>,
+    pub pci_pci_bridges: BTreeMap<String, PciPciBridge>,
```

Comment:
I tried it both ways and went with the type alias for simplicity. Specs don't (yet) have naming restrictions like Omicron's, so they don't need a special `TryFrom` implementation to reject invalid names; it seemed to me that in that scenario, just using the type alias avoided a lot of conversion/Deref/etc. boilerplate that it seemed like I'd otherwise need.